wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -306,199 +372,246 @@ void publishEvent(EventType type, String workspaceId, String error) {
                 LOG.error(format("Could not destroy machine '%s' of workspace '%s'",
                                  machine.getId(),
                                  machine.getWorkspaceId()),
                           ex);
             }
         }
-        // destroying dev-machine
+
+        // Synchronously destroying dev-machine
         try {
             machineManager.destroy(devMachine.getId(), false);
-            publishEvent(EventType.STOPPED, wsId, null);
+            publishEvent(EventType.STOPPED, workspaceId, null);
         } catch (NotFoundException ignore) {
-            // it is ok, machine has been already destroyed
+            // This may happen, if machine is stopped by direct call to the Machine API
+            // MachineManager cleanups all the machines due to application server shutdown
+            // In this case workspace is considered as successfully stopped
+            publishEvent(EventType.STOPPED, workspaceId, null);
         } catch (RuntimeException | ServerException ex) {
-            publishEvent(EventType.ERROR, wsId, ex.getLocalizedMessage());
+            publishEvent(EventType.ERROR, workspaceId, ex.getLocalizedMessage());
             throw ex;
         } finally {
-            removeRuntime(wsId);
+            removeRuntime(workspaceId);
         }
     }
 
-    private void startQueue(String wsId, String envName, boolean recover) throws ServerException,
-                                                                                 NotFoundException,
-                                                                                 ConflictException {
-        publishEvent(EventType.STARTING, wsId, null);
-        MachineConfigImpl config = getPeekConfig(wsId);
+    private void startQueue(String workspaceId,
+                            String envName,
+                            boolean recover) throws ServerException,
+                                                    NotFoundException,
+                                                    ConflictException {
+        publishEvent(EventType.STARTING, workspaceId, null);
+
+        // Starting all the machines one by one by getting configs
+        // from the corresponding starting queue.
+        // Config will be null only if there are no machines left in the queue
+        MachineConfigImpl config = queuePeekOrFail(workspaceId);
         while (config != null) {
-            startMachine(config, wsId, envName, recover);
-            config = getPeekConfig(wsId);
-        }
 
-        // Clean up the start queue when all the machines successfully started
-        rwLock.writeLock().lock();
-        try {
-            final Queue<MachineConfigImpl> queue = startQueues.get(wsId);
-            if (queue != null && queue.isEmpty()) {
-                startQueues.remove(wsId);
+            // According to WorkspaceStatus specification the workspace start
+            // is failed when dev-machine start is failed, so if any error
```

Comment:
Well it is dev-machine throughout all the code and [model](https://github.com/eclipse/che/blob/master/core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/MachineConfig.java#L36), so as long as model is not changed it is dev-machine.